tool/teleport/common/teleport.go

```diff
@@ -469,6 +491,163 @@ func OnStart(clf config.CommandLineFlags, config *service.Config) error {
 }
 
+// GenerateKeys generates TLS and SSH keypairs.
+func GenerateKeys() (private, sshpub, tlspub []byte, err error) {
+	privateKey, publicKey, err := native.GenerateKeyPair()
+	if err != nil {
+		return nil, nil, nil, trace.Wrap(err)
+	}
+
+	sshPrivateKey, err := ssh.ParseRawPrivateKey(privateKey)
+	if err != nil {
+		return nil, nil, nil, trace.Wrap(err)
+	}
+
+	tlsPublicKey, err := tlsca.MarshalPublicKeyFromPrivateKeyPEM(sshPrivateKey)
+	if err != nil {
+		return nil, nil, nil, trace.Wrap(err)
+	}
+
+	return privateKey, publicKey, tlsPublicKey, nil
+}
+
+func onJoin(clf config.CommandLineFlags) error {
+	addr, err := utils.ParseAddr(clf.ProxyServer)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	privateKey, sshPublicKey, tlsPublicKey, err := GenerateKeys()
+	if err != nil {
+		return trace.Wrap(err, "unable to generate new keypairs")
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// TODO(amk) get uuid from a cli argument once agentless inventory management is implemented to allow tsh ssh access via uuid
+	uuid, err := uuid.GenerateUUID()
+	_ = err
+	principals := append(strings.Split(clf.AdditionalPrincipals, ","), uuid)
+
+	certs, err := auth.Register(
+		auth.RegisterParams{
+			Token:                clf.AuthToken,
+			AdditionalPrincipals: principals,
+			JoinMethod:           types.JoinMethod(clf.JoinMethod),
+			ID: auth.IdentityID{
+				Role:     types.RoleNode,
+				NodeName: hostname,
+				HostUUID: uuid,
+			},
+			ProxyServer:        *addr,
+			PublicTLSKey:       tlsPublicKey,
+			PublicSSHKey:       sshPublicKey,
+			GetHostCredentials: client.HostCredentials,
+		},
+	)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	fmt.Printf("Writing Teleport keys to %s\n", clf.OpenSSHKeysPath)
+	if err := writeKeys(clf.OpenSSHKeysPath, privateKey, certs); err != nil {
+		return trace.Wrap(err)
+	}
+
+	fmt.Println("Updating OpenSSH config")
+	if err := updateSSHDConfig(clf.OpenSSHKeysPath, clf.OpenSSHConfigPath); err != nil {
+		return trace.Wrap(err)
+	}
+
+	fmt.Println("Restarting the OpenSSH daemon")
+	if err := restartSSHD(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func writeKeys(sshdConfigDir string, private []byte, certs *proto.Certs) error {
+	if err := os.WriteFile(filepath.Join(sshdConfigDir, "teleport"), private, 0600); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := os.WriteFile(filepath.Join(sshdConfigDir, "teleport-cert.pub"), certs.SSH, 0600); err != nil {
+		return trace.Wrap(err)
+	}
+
+	certsFile, err := os.Create(filepath.Join(sshdConfigDir, "teleport_user_ca.pub"))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer certsFile.Close()
+	for _, cert := range certs.SSHUserCACerts {
+		if _, err := certsFile.Write(cert); err != nil {
+			return trace.Wrap(err)
+		}
+		if _, err := certsFile.WriteString("\n"); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+func updateSSHDConfig(keyDir, sshdConfigPath string) error {
```

Comment:
nit: do we care to check if sshd config is already updated? it's already checked in the shell script so probably not a big deal.